### PR TITLE
Remove S param and use XPure for purely sync effects

### PR DESCRIPTION
**Pull request by mikearnaldi**
Following #291 

- [x] add XPure to Effect primitives & fiber
- [x] remove "S" everywhere

### CODE REVIEW

File: packages/system/src/Fiber/context.ts [hunk 1]
```diff
@@ -776,6 +777,20 @@ export class FiberContext<E, A> implements Fiber.Runtime<E, A> {
                   break
                 }
 
+                case "XPure": {
+                  const res: E.Either<any, any> = X.runEither(
+                    X.provideAll(this.environments?.value || {})(current)
+                  )
+
+                  if (res._tag === "Left") {
+                    current = fail.fail(res)[_I]
```
Comment (qlonik) [inline packages/system/src/Fiber/context.ts:786]: should this be `res.left` passed to `fail.fail`?

Comment (mikearnaldi) [reply packages/system/src/Fiber/context.ts:786]: yes

File: packages/system/src/Effect/catchAllCause_.ts [hunk 2]
```diff
@@ -5,18 +5,18 @@ import type { Effect } from "./effect"
...
  * Recovers from all errors with provided cause.
  */
-export function catchAllCause_<S2, R2, E2, A2, S, R, E, A>(
-  effect: Effect<S2, R2, E2, A2>,
-  f: (_: Cause<E2>) => Effect<S, R, E, A>
+export function catchAllCause_<R2, E2, A2, R, E, A>(
+  effect: Effect<R2, E2, A2>,
+  f: (_: Cause<E2>) => Effect<R, E, A>
 ) {
   return foldCauseM_(effect, f, (x) => succeed(x))
```
Comment (ssledorze) [inline packages/system/src/Effect/catchAllCause_.ts:12]: eta reduction opportunity

File: packages/system/src/Effect/catchAll.ts [hunk 3]
```diff
@@ -5,7 +5,6 @@ import { foldM_ } from "./foldM_"
 /**
  * Recovers from all errors.
  */
-export function catchAll<S, R, E, E2, A>(f: (e: E2) => Effect<S, R, E, A>) {
-  return <S2, R2, A2>(effect: Effect<S2, R2, E2, A2>) =>
-    foldM_(effect, f, (x) => succeed(x))
+export function catchAll<R, E, E2, A>(f: (e: E2) => Effect<R, E, A>) {
+  return <R2, A2>(effect: Effect<R2, E2, A2>) => foldM_(effect, f, (x) => succeed(x))
```
Comment (ssledorze) [inline packages/system/src/Effect/catchAll.ts:9]: eta reduction opportunity

File: packages/system/src/Effect/catchAllCause_.ts [hunk 4]
```diff
@@ -5,18 +5,18 @@ import type { Effect } from "./effect"
...
 /**
  * Recovers from all errors with provided cause.
  */
-export function catchAllCause<S2, R2, E2, A2, S, R, E, A>(
-  f: (_: Cause<E2>) => Effect<S, R, E, A>
+export function catchAllCause<R2, E2, A2, R, E, A>(
+  f: (_: Cause<E2>) => Effect<R, E, A>
 ) {
-  return (effect: Effect<S2, R2, E2, A2>) => foldCauseM_(effect, f, (x) => succeed(x))
+  return (effect: Effect<R2, E2, A2>) => foldCauseM_(effect, f, (x) => succeed(x))
```
Comment (ssledorze) [inline packages/system/src/Effect/catchAllCause_.ts:21]: eta reduction opportunity

File: packages/system/src/Effect/catchAll_.ts [hunk 5]
```diff
@@ -5,9 +5,9 @@ import { foldM_ } from "./foldM_"
...
  * Recovers from all errors.
  */
-export function catchAll_<S2, R2, E2, A2, S, R, E, A>(
-  effect: Effect<S2, R2, E2, A2>,
-  f: (e: E2) => Effect<S, R, E, A>
+export function catchAll_<R2, E2, A2, R, E, A>(
+  effect: Effect<R2, E2, A2>,
+  f: (e: E2) => Effect<R, E, A>
 ) {
   return foldM_(effect, f, (x) => succeed(x))
```
Comment (ssledorze) [inline packages/system/src/Effect/catchAll_.ts:12]: eta reduction opportunity

**Review by ssledorze — APPROVED** [review]
Pretty mechanical (but the introduction of Sync).
Obviously; overall looks cleaner / simpler.
Good to go.

P.S.: Maybe when transitioning RWC; some Sync functions can be required (sync bracket for instance..).

**Comment by mikearnaldi** [comment]
> Pretty mechanical (but the introduction of Sync).
> Obviously; overall looks cleaner / simpler.
> Good to go.
> 
> P.S.: Maybe when transitioning RWC; some Sync functions can be required (sync bracket for instance..).

Correct, more functions might be needed bracket being one, I was also thinking to add a small Managed module based on Sync, do you think it might be needed? I am mixed because in reality when I get to use bracket or Managed its always in async ops like database opening

File: packages/system/src/Stream/Stream/done.ts [hunk 6]
```diff
@@ -1,9 +1,9 @@
 import { done as _ } from "../../Effect"
 import type { Exit } from "../../Exit"
-import type { SyncE } from "./definitions"
+import type { IO } from "./definitions"
 import { fromEffect } from "./fromEffect"
 
 /**
- * The stream that ends with the {@link Exit} value `exit`.
+ * The stream that ends with the `Exit` value `exit`.
```
Comment (qlonik) [inline packages/system/src/Stream/Stream/done.ts:7]: why not link to Exit in jsdoc?

**Review by qlonik — APPROVED** [review]
Looks good to me too. I also added bunch of comments on unrelated things because I just noticed them. Ignore if they dont make sense.

File: packages/system/src/Stream/Stream/done.ts (hunk 6 above)
Comment (mikearnaldi) [reply packages/system/src/Stream/Stream/done.ts:7]: links are not properly shown in vscode highlight no idea why....

Screenshot 2020-09-30 114449

Comment (qlonik) [reply packages/system/src/Stream/Stream/done.ts:7]: oh, i see. but I guess the benefit is that they allow jumping to the Exit interface with ctrl-click

Comment (mikearnaldi) [reply packages/system/src/Stream/Stream/done.ts:7]: Jumping how?

File: packages/system/src/Effect/scope.ts [hunk 7]
```diff
@@ -4,17 +4,17 @@ import type { Runtime } from "../Fiber/core"
 import * as O from "../Option"
 import type { Scope } from "../Scope"
 import { globalScope } from "../Scope"
-import type { AsyncR, Effect } from "./effect"
+import type { Effect, RIO } from "./effect"
 import { IFork, IGetForkScope, IRaceWith } from "./primitives"
```
[the remark this answers is not in the document]
Comment (mikearnaldi) [reply packages/system/src/Effect/scope.ts]: generates circular dependencies between scope & core in constants, moved all scope related to "core-scope.ts"

File: packages/system/src/Effect/catchSome.ts [hunk 8]
```diff
@@ -8,22 +8,22 @@ import type { Effect } from "./effect"
...
-  f: (e: E) => O.Option<Effect<S2, R2, E2, A2>>
+export function catchSome_<R, E, A, R2, E2, A2>(
+  fa: Effect<R, E, A>,
+  f: (e: E) => O.Option<Effect<R2, E2, A2>>
 ) {
   return foldCauseM_(
     fa,
-    (cause): Effect<S2, R2, E | E2, A2> =>
+    (cause): Effect<R2, E | E2, A2> =>
       pipe(
```
[the remark this answers is not in the document]
Comment (mikearnaldi) [reply packages/system/src/Effect/catchSome.ts:27]: "cause" is used in the body so flow is not an option here

File: packages/system/src/Stream/Stream/done.ts (hunk 6 above)
Comment (qlonik) [reply packages/system/src/Stream/Stream/done.ts:7]: IntelliJ allows jumping to definition, when you ctrl-click on the interface/class wrapped into `{@link }` tag

**Pull request merged**